File: crates/multipathd-structs/src/path.rs
use crate::{
    common::{DeviceState, MarginalState, PathDeviceMapperState},
    deserializers::{hex_string, optional_string},
};
use serde::Deserialize;
use strum_macros::{Display, EnumIter};

// NOTE(mnaser): This was retrieved from the following, in case we missed any:
//               https://github.com/opensvc/multipath-tools/blob/master/libmultipath/checkers.h#L84-L91
#[derive(Debug, Default, PartialEq, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum PathChecker {
    #[serde(rename = "directio")]
    DirectIO,

    #[serde(rename = "tur")]
    TestUnitReady,

    #[serde(rename = "hp_sw")]
    HpServiceGuard,

    #[serde(rename = "rdac")]
    RDAC,

    #[serde(rename = "emc_clariion")]
    EmcClariion,

    #[serde(rename = "readsector0")]
    ReadSector0,

    #[serde(rename = "cciss_tur")]
    HpSmartArrayTestUnitReady,

    #[serde(rename = "none")]
    None,

    #[default]
    #[serde(other)]
    Invalid,
}

// NOTE(mnaser): This was retrieved from the following, in case we missed any:
//               https://github.com/opensvc/multipath-tools/blob/master/libmultipath/print.c#L523-L547
#[derive(Clone, Debug, Default, Display, PartialEq, Deserialize, EnumIter)]
#[serde(rename_all = "lowercase")]
pub enum PathCheckerState {
    Ready,
    Faulty,
    Shaky,
    Ghost,
    Delayed,

    #[serde(rename = "i/o pending")]
    IoPending,

    #[serde(rename = "i/o timeout")]
    IoTimeout,

    #[default]
    #[serde(rename = "undef")]
    Undefined,
}

#[derive(Debug, Default, PartialEq, Deserialize)]
#[serde(default)]
pub struct Path {
    #[serde(rename = "dev")]
    pub device_name: String,

    #[serde(rename = "dev_t")]
    pub device_major_minor: String,

    #[serde(rename = "dm_st")]
    pub device_mapper_state: PathDeviceMapperState,

    #[serde(rename = "dev_st")]
    pub device_state: DeviceState,

    #[serde(rename = "checker")]
    pub checker: PathChecker,

    #[serde(rename = "chk_st")]
    pub checker_state: PathCheckerState,

    #[serde(rename = "pri")]
    pub priority: u64,

    #[serde(rename = "host_wwnn")]
    #[serde(deserialize_with = "optional_string")]
    pub host_world_wide_node_name: Option<String>,

    #[serde(rename = "target_wwnn")]
    pub target_world_wide_node_name: String,

    #[serde(rename = "host_wwpn")]
    #[serde(deserialize_with = "optional_string")]
    pub host_world_wide_port_name: Option<String>,

    #[serde(rename = "target_wwpn")]
    #[serde(deserialize_with = "optional_string")]
    pub target_world_wide_port_name: Option<String>,

    #[serde(deserialize_with = "optional_string")]
    pub host_adapter: Option<String>,

    #[serde(rename = "lun_hex")]
    #[serde(deserialize_with = "hex_string")]
    pub logical_unit_number: Option<u64>,

    #[serde(rename = "marginal_st")]
    pub marginal_path_state: MarginalState,
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::common::MarginalState;
    use pretty_assertions::assert_eq;
    use rstest::rstest;
    use serde_json::json;

    #[rstest]
    #[case(json!({
        "dev" : "sdag",
        "dev_t" : "66:0",
        "dm_st" : "active",
        "dev_st" : "running",
        "chk_st" : "ready",
        "checker" : "tur",
        "pri" : 50,
        "host_wwnn" : "[undef]",
        "target_wwnn" : "iqn.2010-06.com.purestorage:flasharray.1007a309fb52d942",
        "host_wwpn" : "[undef]",
        "target_wwpn" : "[undef]",
        "host_adapter" : "192.168.131.31",
        "lun_hex" : "0x000e000000000000",
        "marginal_st" : "normal"
     }), Path {
        device_name: "sdag".into(),
        device_major_minor: "66:0".into(),
        device_mapper_state: PathDeviceMapperState::Active,
        device_state: DeviceState::Running,
        checker_state: PathCheckerState::Ready,
        checker: PathChecker::TestUnitReady,
        priority: 50,
        host_world_wide_node_name: None,
        target_world_wide_node_name: "iqn.2010-06.com.purestorage:flasharray.1007a309fb52d942"
            .into(),
        host_world_wide_port_name: None,
        target_world_wide_port_name: None,
        host_adapter: Some("192.168.131.31".into()),
        logical_unit_number: Some(0x000e000000000000),
        marginal_path_state: MarginalState::Normal,
    })]
    #[case(json!({
        "dev" : "sdcx",
        "dev_t" : "70:80",
        "dm_st" : "active",
        "dev_st" : "running",
        "chk_st" : "ready",
        "checker" : "tur",
        "pri" : 50,
        "host_wwnn" : "0x20000090fa678144",
        "target_wwnn" : "0x524a937247e74300",
        "host_wwpn" : "0x10000090fa678144",
        "target_wwpn" : "0x524a937247e74300",
        "host_adapter" : "0000:00:03.0",
        "marginal_st" : "normal"
     }), Path {
        device_name: "sdcx".into(),
        device_major_minor: "70:80".into(),
        device_mapper_state: PathDeviceMapperState::Active,
        device_state: DeviceState::Running,
        checker_state: PathCheckerState::Ready,
        checker: PathChecker::TestUnitReady,
        priority: 50,
        host_world_wide_node_name: Some("0x20000090fa678144".into()),
        target_world_wide_node_name: "0x524a937247e74300".into(),
        host_world_wide_port_name: Some("0x10000090fa678144".into()),
        target_world_wide_port_name: Some("0x524a937247e74300".into()),
        host_adapter: Some("0000:00:03.0".into()),
        marginal_path_state: MarginalState::Normal,
        ..Default::default()
    })]
    fn test_parsing_path(#[case] input: serde_json::Value, #[case] expected: Path) {
        let result: Path = serde_json::from_value(input).unwrap();
        assert_eq!(result, expected);
    }
}

File: internal/multipath/path.go
//go:generate go-enum -f=$GOFILE --marshal --names --values

// Copyright (c) 2024 VEXXHOST, Inc.
// SPDX-License-Identifier: Apache-2.0

package multipath

import (
	"encoding/json"
)

// NOTE(mnaser): This was retrieved from the following, in case we missed any:
//
//	https://github.com/opensvc/multipath-tools/blob/master/libmultipath/checkers.h#L84-L91
type PathChecker string

const (
	PathCheckerDirectIO                  PathChecker = "directio"
	PathCheckerTestUnitReady             PathChecker = "tur"
	PathCheckerHpServiceGuard            PathChecker = "hp_sw"
	PathCheckerRDAC                      PathChecker = "rdac"
	PathCheckerEmcClariion               PathChecker = "emc_clariion"
	PathCheckerReadSector0               PathChecker = "readsector0"
	PathCheckerHpSmartArrayTestUnitReady PathChecker = "cciss_tur"
	PathCheckerNone                      PathChecker = "none"
	PathCheckerInvalid                   PathChecker = "invalid"
)

// NOTE(mnaser): This was retrieved from the following, in case we missed any:
//
//	https://github.com/opensvc/multipath-tools/blob/master/libmultipath/print.c#L523-L547
//
// ENUM(ready, faulty, shaky, ghost, delayed, i/o pending, i/o timeout, undef)
type PathCheckerState string

// Path represents a multipath device path
type Path struct {
	DeviceName              string                `json:"dev"`
	DeviceMajorMinor        string                `json:"dev_t"`
	DeviceMapperState       PathDeviceMapperState `json:"dm_st"`
	DeviceState             DeviceState           `json:"dev_st"`
	Checker                 PathChecker           `json:"checker"`
	CheckerState            PathCheckerState      `json:"chk_st"`
	Priority                uint64                `json:"pri"`
	HostWorldWideNodeName   *string               `json:"host_wwnn"`
	TargetWorldWideNodeName string                `json:"target_wwnn"`
	HostWorldWidePortName   *string               `json:"host_wwpn"`
	TargetWorldWidePortName *string               `json:"target_wwpn"`
	HostAdapter             *string               `json:"host_adapter"`
	LogicalUnitNumber       *uint64               `json:"lun_hex"`
	MarginalPathState       MarginalState         `json:"marginal_st"`
}

// UnmarshalJSON implements custom JSON unmarshaling for Path
func (p *Path) UnmarshalJSON(data []byte) error {
	type Alias Path
	aux := &struct {
		HostWorldWideNodeName   interface{} `json:"host_wwnn"`
		HostWorldWidePortName   interface{} `json:"host_wwpn"`
		TargetWorldWidePortName interface{} `json:"target_wwpn"`
		HostAdapter             interface{} `json:"host_adapter"`
		LogicalUnitNumber       interface{} `json:"lun_hex"`
		*Alias
	}{
		Alias: (*Alias)(p),
	}

	if err := json.Unmarshal(data, &aux); err != nil {
		return err
	}

	p.HostWorldWideNodeName = parseOptionalString(aux.HostWorldWideNodeName)
	p.HostWorldWidePortName = parseOptionalString(aux.HostWorldWidePortName)
	p.TargetWorldWidePortName = parseOptionalString(aux.TargetWorldWidePortName)
	p.HostAdapter = parseOptionalString(aux.HostAdapter)

	if aux.LogicalUnitNumber != nil {
		if hexStr, ok := aux.LogicalUnitNumber.(string); ok {
			p.LogicalUnitNumber = parseHexString(hexStr)
		}
	}

	return nil
}

// parseOptionalString converts interface{} to *string, treating "[undef]" as nil
func parseOptionalString(v interface{}) *string {
	if v == nil {
		return nil
	}

	str, ok := v.(string)
	if !ok || str == "[undef]" || str == "" {
		return nil
	}

	return &str
}
